File: src/Auth/Crypto/Totp.hs
module Auth.Crypto.Totp
  ( generateSecret
  , secretToBase32
  , otpauthUri
  , verifyCode
  ) where

import           Crypto.Hash.Algorithms (SHA1 (..))
import           Crypto.MAC.HMAC        (HMAC, hmac)
import           Crypto.Random          (getRandomBytes)
import           Data.Bits              (shiftL, shiftR, (.&.), (.|.))
import           Data.ByteArray         (convert)
import qualified Data.ByteString        as BS
import           Data.Text              (Text)
import qualified Data.Text              as T
import           Data.Time.Clock.POSIX  (getPOSIXTime)
import           Data.Word              (Word32, Word64)
import           Text.Printf            (printf)

import qualified Auth.Crypto.Base32     as B32

period :: Word64
period = 30

digits :: Int
digits = 6

generateSecret :: IO BS.ByteString
generateSecret = getRandomBytes 20

secretToBase32 :: BS.ByteString -> Text
secretToBase32 = T.pack . B32.encode

otpauthUri :: Text -> Text -> BS.ByteString -> Text
otpauthUri issuer account secret =
  "otpauth://totp/" <> enc issuer <> ":" <> enc account
    <> "?secret=" <> secretToBase32 secret
    <> "&issuer=" <> enc issuer
    <> "&algorithm=SHA1&digits=" <> T.pack (show digits)
    <> "&period=" <> T.pack (show period)
  where
    enc = T.concatMap escape
    escape ' ' = "%20"
    escape '/' = "%2F"
    escape ':' = "%3A"
    escape '?' = "%3F"
    escape '&' = "%26"
    escape '=' = "%3D"
    escape c   = T.singleton c

counterBytes :: Word64 -> BS.ByteString
counterBytes c = BS.pack [ fromIntegral (c `shiftR` (8 * (7 - i)) .&. 0xff) | i <- [0 .. 7] ]

hotp :: BS.ByteString -> Word64 -> String
hotp secret counter =
  let mac    = convert (hmac secret (counterBytes counter) :: HMAC SHA1) :: BS.ByteString
      offset = fromIntegral (BS.index mac 19 .&. 0x0f)
      at i   = fromIntegral (BS.index mac (offset + i)) :: Word32
      binCode = ((at 0 .&. 0x7f) `shiftL` 24)
            .|. ((at 1 .&. 0xff) `shiftL` 16)
            .|. ((at 2 .&. 0xff) `shiftL` 8)
            .|.  (at 3 .&. 0xff)
      value  = binCode `mod` (10 ^ digits)
  in printf "%0*u" digits value

verifyCode :: BS.ByteString -> Text -> IO Bool
verifyCode secret code = do
  now <- getPOSIXTime
  let counter   = floor (realToFrac now / fromIntegral period :: Double) :: Word64
      candidate = T.unpack (T.strip code)
      valid     = [ hotp secret c | c <- [counter - 1, counter, counter + 1] ]
  pure (candidate `elem` valid)

File: src/api/auth/crypto/totp.rs
use std::time::{SystemTime, UNIX_EPOCH};

use base32::Alphabet;
use hmac::{Hmac, Mac};
use sha1::Sha1;

const PERIOD: u64 = 30;
const DIGITS: u32 = 6;
const SECRET_LENGTH: usize = 20;

pub fn generate_secret() -> Vec<u8> {
    super::token::random_bytes(SECRET_LENGTH)
}

pub fn secret_to_base32(secret: &[u8]) -> String {
    base32::encode(Alphabet::Rfc4648 { padding: false }, secret)
}

pub fn otpauth_uri(issuer: &str, account: &str, secret: &[u8]) -> String {
    let enc = |s: &str| -> String {
        s.chars()
            .map(|c| match c {
                ' ' => "%20".to_string(),
                '/' => "%2F".to_string(),
                ':' => "%3A".to_string(),
                '?' => "%3F".to_string(),
                '&' => "%26".to_string(),
                '=' => "%3D".to_string(),
                other => other.to_string(),
            })
            .collect()
    };
    format!(
        "otpauth://totp/{issuer}:{account}?secret={secret}&issuer={issuer}&algorithm=SHA1&digits={digits}&period={period}",
        issuer = enc(issuer),
        account = enc(account),
        secret = secret_to_base32(secret),
        digits = DIGITS,
        period = PERIOD,
    )
}

fn hotp(secret: &[u8], counter: u64) -> String {
    let mut mac = Hmac::<Sha1>::new_from_slice(secret).expect("HMAC accepts keys of any length");
    mac.update(&counter.to_be_bytes());
    let digest = mac.finalize().into_bytes();

    let offset = (digest[19] & 0x0f) as usize;
    let bin_code = ((u32::from(digest[offset]) & 0x7f) << 24)
        | ((u32::from(digest[offset + 1]) & 0xff) << 16)
        | ((u32::from(digest[offset + 2]) & 0xff) << 8)
        | (u32::from(digest[offset + 3]) & 0xff);
    let value = bin_code % 10u32.pow(DIGITS);
    format!("{value:0width$}", width = DIGITS as usize)
}

pub fn verify_code(secret: &[u8], code: &str) -> bool {
    let now = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_secs())
        .unwrap_or(0);
    let counter = now / PERIOD;
    let candidate = code.trim();
    [counter.wrapping_sub(1), counter, counter + 1]
        .iter()
        .any(|c| hotp(secret, *c) == candidate)
}
